src/dash-table/components/CellFactory.tsx

```diff
@@ -111,35 +112,65 @@ export default class CellFactory {
             dropdowns
         );
 
-        const ops = matrixMap2(
+        const ops = this.getDataOpCells(
             operations,
             dataOpStyles,
-            (o, s, i, j) => React.cloneElement(o, {
-                style: R.mergeAll([
-                    dataOpEdges && dataOpEdges.getStyle(i, j),
-                    s,
-                    o.props.style
-                ])
-            })
+            dataOpEdges
         );
 
-        const cells = matrixMap3(
+        const cells = this.getDataCells(
             cellWrappers,
-            cellStyles,
             cellContents,
-            (w, s, c, i, j) => React.cloneElement(w, {
-                children: [c],
-                style: R.mergeAll([
-                    s,
-                    dataEdges && dataEdges.getStyle(i, j)
-                ])
-            })
+            cellStyles,
+            dataEdges
         );
 
-        return arrayMap2(
+        return this.getCells(
             ops,
-            cells,
-            (o, c) => Array.prototype.concat(o, c)
+            cells
         );
     }
+
+    getCells = memoizeOne((
+        opCells: JSX.Element[][],
+        dataCells: JSX.Element[][]
+    ) => arrayMap2(
+        opCells,
+        dataCells,
+        (o, c) => [...o, ...c]
```

Comment:
no opCells must be a very common case - could we detect it here and avoid this `arrayMap2` altogether?

Reply:
Sure. Out of curiosity, checked out the speed of various ways of doing this
![image](https://user-images.githubusercontent.com/8092993/59309855-b279d500-8c72-11e9-9fa3-14c679ec698a.png)

Only Chrome shown above, but the spread operator is consistently 2.5-4x slower than `concat`

https://jsperf.com/concat-some-arrays

Reply:
weird... we're still transpiling this, right? I wonder what this expression transpiles into? Maybe Chrome just hasn't bothered optimizing spread yet since presumably most people are still transpiling it.

Reply:
The above example is native Chrome support, not what the table outputs. Spread and destructuring operators are not particularly efficient as they require more checks under the hood and have not been optimized over two decades. In most cases this is not a problem..

In the table's case, we transpile with Babel for the lowest common denominator: IE11 -- which has absolutely no ES6 support. `core-js` is used to polyfill a lot of features but this is pure Babel transpilation.

```
return [...o, ...c]; 
```

becomes:

```
function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

return _toConsumableArray(o).concat(_toConsumableArray(c));
```